File: hs/robot-transfer.hs
-- https://www.codewars.com/kata/58aaa3ca821a767300000017/train/javascript

module RobotTransfer.Kata (robotTransfer) where

import Test.Hspec

robotTransfer :: [[String]] -> Int -> Int
robotTransfer = undefined

spec :: Spec
spec = do
  describe "Robot Transfer" $ do
    it "Example Tests" $ do
      robotTransfer
        [ ["0,1", "0,0", "1,2"],
          ["1,1", "1,0", "0,2"],
          ["2,1", "2,0", "0,0"]
        ]
        2
        `shouldBe` 8
          robotTransfer
          [ ["0,1", "0,0"],
            ["1,1", "1,0"]
          ]
          2
        `shouldBe` 4

File: js/robot-transfer.js
// https://www.codewars.com/kata/58aaa3ca821a767300000017/train/javascript

function finalPosition(matrix, steps, position) {
  console.log(position);
}

function startingPositions(matrix) {
  if (!matrix) {
    return [];
  }
  return [
    ...startingPositions(matrix.slice(1)),
    typeof matrix[0] === "object" ? startingPositions(type) : 1,
  ];
}

function robotTransfer(matrix, steps) {
  return startingPositions(matrix).filter(
    finalPosition(matrix, steps, [x, y]) === [x, y]
  ).length;
}

console.log(
  robotTransfer(
    [
      ["0,1", "0,0", "1,2"],
      ["1,1", "1,0", "0,2"],
      ["2,1", "2,0", "0,0"],
    ],
    2
  ) == 8
);

console.log(
  robotTransfer(
    [
      ["0,1", "0,0"],
      ["1,1", "1,0"],
    ],
    2
  ) == 4
);
